lib/WorkerHandler.js

```diff
@@ -289,11 +294,35 @@ WorkerHandler.prototype.terminate = function (force) {
     }
     this.terminating = false;
     this.terminated = true;
+    if (this.terminationHandler) {
+      this.terminationHandler(this);
```

Comment:
I don't fully understand how the (async?) callback is handled. As far as I can see the only place it's called is here, and this is called synchronously, which would make the callback redundant.

Reply:
1. It is called, of course, only if one has been passed in to the original Workerhandler.terminate call.
2. It isn't redundant, because that part of the code is only executed if/when the process is not busy. If the process is busy, that code is not executed, and the 'terminating' flag is set. So if the process is busy, the callback is not immediately called.
3. In the busy case, the terminate function is called again when the process finally finishes - this is done by the 'on message' handler that is set up n the WorkerHandler constructor. In that second terminate call, the process will not be busy, and at that point the code above will be executed, and the callback is called at that point.
4. Now that I look at it to explain it, maybe we should also call the termination handler if the process errors out (in 'on error'.)

Let me know if it is still not clear.

Reply:
And of course, if the process is not busy, the callback is called immediately, (and the Promise in TerminateAndNotify is resolved immediately,) which is what you want.

Reply:
Bottom line is that it works in my complex case - without it the process would never shut down properly unless I put in an artificial timeout, because I had several child processes that were still busy. With this, everything terminates nicely. I was trying to replicate the situation in a simple unit test, but I haven't been able to do that so far.

Reply:
I should maybe add that our scenario is node.js only - it's an Electron app - so all my testing is done there, and using node child processes. I have no way to test this in the browser case, but it seems it should be ok there as well.

Reply:
Thanks for the updates. I now see what I was missing: `WorkerHandler.terminate` is _first_ called via the `terminateAndNotify`. When not busy, it immeditately calls the callback function. When busy, `terminate` will be called _again_ after the task finished, but now _without_ a new callback function as argument, in which case the callback function (cached as `this.terminationHandler`) is called. 

Maybe the flow becomse a little bit more clear when moving setting of the `terminationHandler` from `terminate()` to `terminateAndNotify()`, since the callback is only passed when calling `terminateAndNotify`?

```js
  if (typeof callback === 'function') {
    this.terminationHandler = callback;
  }
```

Or give a promise based solution another try: on construction of `WorkerHandler`, create a promise
```js
this.terminated = Promise.defer();
```
this promise is returned by `terminateAndNotify`:
```js
WorkerHandler.prototype.terminateAndNotify = function (force, timeout) {
  // ...
  return this.terminated;
}
```
And resolve the promise as soon as the worker terminates or errors:
```js
this.terminated.resolve(this);
```